Cython/Compiler/Nodes.py

```diff
@@ -8937,6 +8937,129 @@ def generate_execution_code(self, code):
         code.put_ensure_gil(declare_gilstate=False)
 
 
+class CriticalSectionStatNode(TryFinallyStatNode):
+    """
+    Represents a freethreading Python critical section.
+    In non-freethreading Python, this is a no-op.
+
+    args    list of ExprNode    1 or 2 elements, must be object
+    """
+
+    child_attrs = ["args"] + TryFinallyStatNode.child_attrs
+
+    # loosely follow the naming convention that Python itself follows
+    DEFAULT_VARIABLE_NAME = "__pyx_cs"
+
+    var_type = None
+    state_temp = None
+
+    def __init__(self, pos, /, args, body, **kwds):
+        if len(args) > 1:
+            self.var_type = PyrexTypes.c_py_critical_section2_type
+        else:
+            self.var_type = PyrexTypes.c_py_critical_section_type
+
+        self.create_state_temp_if_needed(pos, args, body)
+
+        TryFinallyStatNode.__init__(
+            self, pos,
+            args=args,
+            body=body,
+            **kwds,
+            finally_clause=CriticalSectionExitNode(
+                pos, len=len(args), state_temp=self.state_temp))
+
+    def create_state_temp_if_needed(self, pos, args, body):
+        from .ParseTreeTransforms import YieldNodeCollector
+        collector = YieldNodeCollector()
+        collector.visitchildren(body)
+        if not collector.yields:
+            return
+
+        from . import ExprNodes
+        self.state_temp = ExprNodes.TempNode(pos, self.var_type)
+
+    def analyse_declarations(self, env):
+        for arg in self.args:
+            arg.analyse_declarations(env)
+        return super().analyse_declarations(env)
+
+    def analyse_expressions(self, env):
+        env.use_utility_code(
+            UtilityCode.load_cached("CriticalSections", "ModuleSetupCode.c"))
```

Comment:
I think this is too early. It can be included during code generation, when we really know that we need it.